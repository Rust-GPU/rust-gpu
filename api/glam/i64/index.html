<!DOCTYPE html><html lang="en"><head><meta charset="utf-8"><meta name="viewport" content="width=device-width, initial-scale=1.0"><meta name="generator" content="rustdoc"><meta name="description" content="`i64` vector types."><title>glam::i64 - Rust</title><script>if(window.location.protocol!=="file:")document.head.insertAdjacentHTML("beforeend","SourceSerif4-Regular-6b053e98.ttf.woff2,FiraSans-Italic-81dc35de.woff2,FiraSans-Regular-0fe48ade.woff2,FiraSans-MediumItalic-ccf7e434.woff2,FiraSans-Medium-e1aa3f0a.woff2,SourceCodePro-Regular-8badfe75.ttf.woff2,SourceCodePro-Semibold-aa29a496.ttf.woff2".split(",").map(f=>`<link rel="preload" as="font" type="font/woff2" crossorigin href="../../static.files/${f}">`).join(""))</script><link rel="stylesheet" href="../../static.files/normalize-9960930a.css"><link rel="stylesheet" href="../../static.files/rustdoc-84e720fa.css"><meta name="rustdoc-vars" data-root-path="../../" data-static-root-path="../../static.files/" data-current-crate="glam" data-themes="" data-resource-suffix="" data-rustdoc-version="1.89.0-nightly (be19eda0d 2025-06-22)" data-channel="nightly" data-search-js="search-92309212.js" data-settings-js="settings-5514c975.js" ><script src="../../static.files/storage-4e99c027.js"></script><script defer src="../sidebar-items.js"></script><script defer src="../../static.files/main-fd3af306.js"></script><noscript><link rel="stylesheet" href="../../static.files/noscript-32bb7600.css"></noscript><link rel="alternate icon" type="image/png" href="../../static.files/favicon-32x32-6580c154.png"><link rel="icon" type="image/svg+xml" href="../../static.files/favicon-044be391.svg"></head><body class="rustdoc mod"><!--[if lte IE 11]><div class="warning">This old browser is unsupported and will most likely display funky things.</div><![endif]--><nav class="mobile-topbar"><button class="sidebar-menu-toggle" title="show sidebar"></button></nav><nav class="sidebar"><div class="sidebar-crate"><h2><a href="../../glam/index.html">glam</a><span class="version">0.30.2</span></h2></div><div class="sidebar-elems"><section id="rustdoc-toc"><h2 class="location"><a href="#">Module i64</a></h2><h3><a href="#structs">Module Items</a></h3><ul class="block"><li><a href="#structs" title="Structs">Structs</a></li><li><a href="#functions" title="Functions">Functions</a></li></ul></section><div id="rustdoc-modnav"><h2 class="in-crate"><a href="../index.html">In crate glam</a></h2></div></div></nav><div class="sidebar-resizer" title="Drag to resize sidebar"></div><main><div class="width-limiter"><rustdoc-search></rustdoc-search><section id="main-content" class="content"><div class="main-heading"><div class="rustdoc-breadcrumbs"><a href="../index.html">glam</a></div><h1>Module <span>i64</span><button id="copy-path" title="Copy item path to clipboard">Copy item path</button></h1><rustdoc-toolbar></rustdoc-toolbar><span class="sub-heading"><a class="src" href="../../src/glam/i64.rs.html#1-45">Source</a> </span></div><details class="toggle top-doc" open><summary class="hideme"><span>Expand description</span></summary><div class="docblock"><p><code>i64</code> vector types.</p>
</div></details><h2 id="structs" class="section-header">Structs<a href="#structs" class="anchor">ยง</a></h2><dl class="item-table"><dt><a class="struct" href="struct.I64Vec2.html" title="struct glam::i64::I64Vec2">I64Vec2</a></dt><dd>A 2-dimensional vector.</dd><dt><a class="struct" href="struct.I64Vec3.html" title="struct glam::i64::I64Vec3">I64Vec3</a></dt><dd>A 3-dimensional vector.</dd><dt><a class="struct" href="struct.I64Vec4.html" title="struct glam::i64::I64Vec4">I64Vec4</a></dt><dd>A 4-dimensional vector.</dd></dl><h2 id="functions" class="section-header">Functions<a href="#functions" class="anchor">ยง</a></h2><dl class="item-table"><dt><a class="fn" href="fn.i64vec2.html" title="fn glam::i64::i64vec2">i64vec2</a></dt><dd>Creates a 2-dimensional vector.</dd><dt><a class="fn" href="fn.i64vec3.html" title="fn glam::i64::i64vec3">i64vec3</a></dt><dd>Creates a 3-dimensional vector.</dd><dt><a class="fn" href="fn.i64vec4.html" title="fn glam::i64::i64vec4">i64vec4</a></dt><dd>Creates a 4-dimensional vector.</dd></dl></section></div></main></body></html>