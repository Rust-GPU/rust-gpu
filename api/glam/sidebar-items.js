window.SIDEBAR_ITEMS = {"enum":["EulerRot"],"mod":["bool","f32","f64","i16","i32","i64","i8","swizzles","u16","u32","u64","u8"],"trait":["FloatExt"]};