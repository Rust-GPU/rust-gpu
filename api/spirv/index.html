<!DOCTYPE html><html lang="en"><head><meta charset="utf-8"><meta name="viewport" content="width=device-width, initial-scale=1.0"><meta name="generator" content="rustdoc"><meta name="description" content="The SPIR-V header."><title>spirv - Rust</title><script>if(window.location.protocol!=="file:")document.head.insertAdjacentHTML("beforeend","SourceSerif4-Regular-46f98efaafac5295.ttf.woff2,FiraSans-Regular-018c141bf0843ffd.woff2,FiraSans-Medium-8f9a781e4970d388.woff2,SourceCodePro-Regular-562dcc5011b6de7d.ttf.woff2,SourceCodePro-Semibold-d899c5a5c4aeb14a.ttf.woff2".split(",").map(f=>`<link rel="preload" as="font" type="font/woff2" crossorigin href="../static.files/${f}">`).join(""))</script><link rel="stylesheet" href="../static.files/normalize-76eba96aa4d2e634.css"><link rel="stylesheet" href="../static.files/rustdoc-081576b923113409.css"><meta name="rustdoc-vars" data-root-path="../" data-static-root-path="../static.files/" data-current-crate="spirv" data-themes="" data-resource-suffix="" data-rustdoc-version="1.79.0-nightly (244da22fa 2024-04-23)" data-channel="nightly" data-search-js="search-bf21c90c8c1d92b1.js" data-settings-js="settings-4313503d2e1961c2.js" ><script src="../static.files/storage-e32f0c247825364d.js"></script><script defer src="../crates.js"></script><script defer src="../static.files/main-20a3ad099b048cf2.js"></script><noscript><link rel="stylesheet" href="../static.files/noscript-09095024cf37855e.css"></noscript><link rel="alternate icon" type="image/png" href="../static.files/favicon-32x32-422f7d1d52889060.png"><link rel="icon" type="image/svg+xml" href="../static.files/favicon-2c020d218678b618.svg"></head><body class="rustdoc mod crate"><!--[if lte IE 11]><div class="warning">This old browser is unsupported and will most likely display funky things.</div><![endif]--><nav class="mobile-topbar"><button class="sidebar-menu-toggle" title="show sidebar"></button></nav><nav class="sidebar"><div class="sidebar-crate"><h2><a href="../spirv/index.html">spirv</a><span class="version">0.3.0+sdk-1.3.268.0</span></h2></div><div class="sidebar-elems"><ul class="block"><li><a id="all-types" href="all.html">All Items</a></li></ul><section><ul class="block"><li><a href="#structs">Structs</a></li><li><a href="#enums">Enums</a></li><li><a href="#constants">Constants</a></li><li><a href="#types">Type Aliases</a></li></ul></section></div></nav><div class="sidebar-resizer"></div><main><div class="width-limiter"><nav class="sub"><form class="search-form"><span></span><div id="sidebar-button" tabindex="-1"><a href="../spirv/all.html" title="show sidebar"></a></div><input class="search-input" name="search" aria-label="Run search in the documentation" autocomplete="off" spellcheck="false" placeholder="Type ‘S’ or ‘/’ to search, ‘?’ for more options…" type="search"><div id="help-button" tabindex="-1"><a href="../help.html" title="help">?</a></div><div id="settings-menu" tabindex="-1"><a href="../settings.html" title="settings">Settings</a></div></form></nav><section id="main-content" class="content"><div class="main-heading"><h1>Crate <a class="mod" href="#">spirv</a><button id="copy-path" title="Copy item path to clipboard">Copy item path</button></h1><span class="out-of-band"><a class="src" href="../src/spirv/lib.rs.html#1-14">source</a> · <button id="toggle-all-docs" title="collapse all docs">[<span>&#x2212;</span>]</button></span></div><details class="toggle top-doc" open><summary class="hideme"><span>Expand description</span></summary><div class="docblock"><p>The SPIR-V header.</p>
<p>This crate contains Rust definitions of all SPIR-V structs, enums,
and constants.</p>
<p>The version of this crate is the version of SPIR-V it contains.</p>
</div></details><h2 id="structs" class="section-header">Structs<a href="#structs" class="anchor">§</a></h2><ul class="item-table"><li><div class="item-name"><a class="struct" href="struct.CooperativeMatrixOperands.html" title="struct spirv::CooperativeMatrixOperands">CooperativeMatrixOperands</a></div><div class="desc docblock-short">SPIR-V operand kind: <a href="https://www.khronos.org/registry/spir-v/specs/unified1/SPIRV.html#_a_id_cooperative_matrix_operands_a_cooperative_matrix_operands">CooperativeMatrixOperands</a></div></li><li><div class="item-name"><a class="struct" href="struct.FPFastMathMode.html" title="struct spirv::FPFastMathMode">FPFastMathMode</a></div><div class="desc docblock-short">SPIR-V operand kind: <a href="https://www.khronos.org/registry/spir-v/specs/unified1/SPIRV.html#_a_id_fp_fast_math_mode_a_fp_fast_math_mode">FPFastMathMode</a></div></li><li><div class="item-name"><a class="struct" href="struct.FragmentShadingRate.html" title="struct spirv::FragmentShadingRate">FragmentShadingRate</a></div><div class="desc docblock-short">SPIR-V operand kind: <a href="https://www.khronos.org/registry/spir-v/specs/unified1/SPIRV.html#_a_id_fragment_shading_rate_a_fragment_shading_rate">FragmentShadingRate</a></div></li><li><div class="item-name"><a class="struct" href="struct.FunctionControl.html" title="struct spirv::FunctionControl">FunctionControl</a></div><div class="desc docblock-short">SPIR-V operand kind: <a href="https://www.khronos.org/registry/spir-v/specs/unified1/SPIRV.html#_a_id_function_control_a_function_control">FunctionControl</a></div></li><li><div class="item-name"><a class="struct" href="struct.ImageOperands.html" title="struct spirv::ImageOperands">ImageOperands</a></div><div class="desc docblock-short">SPIR-V operand kind: <a href="https://www.khronos.org/registry/spir-v/specs/unified1/SPIRV.html#_a_id_image_operands_a_image_operands">ImageOperands</a></div></li><li><div class="item-name"><a class="struct" href="struct.KernelProfilingInfo.html" title="struct spirv::KernelProfilingInfo">KernelProfilingInfo</a></div><div class="desc docblock-short">SPIR-V operand kind: <a href="https://www.khronos.org/registry/spir-v/specs/unified1/SPIRV.html#_a_id_kernel_profiling_info_a_kernel_profiling_info">KernelProfilingInfo</a></div></li><li><div class="item-name"><a class="struct" href="struct.LoopControl.html" title="struct spirv::LoopControl">LoopControl</a></div><div class="desc docblock-short">SPIR-V operand kind: <a href="https://www.khronos.org/registry/spir-v/specs/unified1/SPIRV.html#_a_id_loop_control_a_loop_control">LoopControl</a></div></li><li><div class="item-name"><a class="struct" href="struct.MemoryAccess.html" title="struct spirv::MemoryAccess">MemoryAccess</a></div><div class="desc docblock-short">SPIR-V operand kind: <a href="https://www.khronos.org/registry/spir-v/specs/unified1/SPIRV.html#_a_id_memory_access_a_memory_access">MemoryAccess</a></div></li><li><div class="item-name"><a class="struct" href="struct.MemorySemantics.html" title="struct spirv::MemorySemantics">MemorySemantics</a></div><div class="desc docblock-short">SPIR-V operand kind: <a href="https://www.khronos.org/registry/spir-v/specs/unified1/SPIRV.html#_a_id_memory_semantics_a_memory_semantics">MemorySemantics</a></div></li><li><div class="item-name"><a class="struct" href="struct.RayFlags.html" title="struct spirv::RayFlags">RayFlags</a></div><div class="desc docblock-short">SPIR-V operand kind: <a href="https://www.khronos.org/registry/spir-v/specs/unified1/SPIRV.html#_a_id_ray_flags_a_ray_flags">RayFlags</a></div></li><li><div class="item-name"><a class="struct" href="struct.SelectionControl.html" title="struct spirv::SelectionControl">SelectionControl</a></div><div class="desc docblock-short">SPIR-V operand kind: <a href="https://www.khronos.org/registry/spir-v/specs/unified1/SPIRV.html#_a_id_selection_control_a_selection_control">SelectionControl</a></div></li></ul><h2 id="enums" class="section-header">Enums<a href="#enums" class="anchor">§</a></h2><ul class="item-table"><li><div class="item-name"><a class="enum" href="enum.AccessQualifier.html" title="enum spirv::AccessQualifier">AccessQualifier</a></div><div class="desc docblock-short">SPIR-V operand kind: <a href="https://www.khronos.org/registry/spir-v/specs/unified1/SPIRV.html#_a_id_access_qualifier_a_access_qualifier">AccessQualifier</a></div></li><li><div class="item-name"><a class="enum" href="enum.AddressingModel.html" title="enum spirv::AddressingModel">AddressingModel</a></div><div class="desc docblock-short">SPIR-V operand kind: <a href="https://www.khronos.org/registry/spir-v/specs/unified1/SPIRV.html#_a_id_addressing_model_a_addressing_model">AddressingModel</a></div></li><li><div class="item-name"><a class="enum" href="enum.BuiltIn.html" title="enum spirv::BuiltIn">BuiltIn</a></div><div class="desc docblock-short">SPIR-V operand kind: <a href="https://www.khronos.org/registry/spir-v/specs/unified1/SPIRV.html#_a_id_built_in_a_built_in">BuiltIn</a></div></li><li><div class="item-name"><a class="enum" href="enum.CLOp.html" title="enum spirv::CLOp">CLOp</a></div><div class="desc docblock-short"><a href="https://www.khronos.org/registry/spir-v/specs/unified1/OpenCL.ExtendedInstructionSet.100.html">OpenCL.std</a> extended instruction opcode</div></li><li><div class="item-name"><a class="enum" href="enum.Capability.html" title="enum spirv::Capability">Capability</a></div><div class="desc docblock-short">SPIR-V operand kind: <a href="https://www.khronos.org/registry/spir-v/specs/unified1/SPIRV.html#_a_id_capability_a_capability">Capability</a></div></li><li><div class="item-name"><a class="enum" href="enum.CooperativeMatrixLayout.html" title="enum spirv::CooperativeMatrixLayout">CooperativeMatrixLayout</a></div><div class="desc docblock-short">SPIR-V operand kind: <a href="https://www.khronos.org/registry/spir-v/specs/unified1/SPIRV.html#_a_id_cooperative_matrix_layout_a_cooperative_matrix_layout">CooperativeMatrixLayout</a></div></li><li><div class="item-name"><a class="enum" href="enum.CooperativeMatrixUse.html" title="enum spirv::CooperativeMatrixUse">CooperativeMatrixUse</a></div><div class="desc docblock-short">SPIR-V operand kind: <a href="https://www.khronos.org/registry/spir-v/specs/unified1/SPIRV.html#_a_id_cooperative_matrix_use_a_cooperative_matrix_use">CooperativeMatrixUse</a></div></li><li><div class="item-name"><a class="enum" href="enum.Decoration.html" title="enum spirv::Decoration">Decoration</a></div><div class="desc docblock-short">SPIR-V operand kind: <a href="https://www.khronos.org/registry/spir-v/specs/unified1/SPIRV.html#_a_id_decoration_a_decoration">Decoration</a></div></li><li><div class="item-name"><a class="enum" href="enum.Dim.html" title="enum spirv::Dim">Dim</a></div><div class="desc docblock-short">SPIR-V operand kind: <a href="https://www.khronos.org/registry/spir-v/specs/unified1/SPIRV.html#_a_id_dim_a_dim">Dim</a></div></li><li><div class="item-name"><a class="enum" href="enum.ExecutionMode.html" title="enum spirv::ExecutionMode">ExecutionMode</a></div><div class="desc docblock-short">SPIR-V operand kind: <a href="https://www.khronos.org/registry/spir-v/specs/unified1/SPIRV.html#_a_id_execution_mode_a_execution_mode">ExecutionMode</a></div></li><li><div class="item-name"><a class="enum" href="enum.ExecutionModel.html" title="enum spirv::ExecutionModel">ExecutionModel</a></div><div class="desc docblock-short">SPIR-V operand kind: <a href="https://www.khronos.org/registry/spir-v/specs/unified1/SPIRV.html#_a_id_execution_model_a_execution_model">ExecutionModel</a></div></li><li><div class="item-name"><a class="enum" href="enum.FPDenormMode.html" title="enum spirv::FPDenormMode">FPDenormMode</a></div><div class="desc docblock-short">SPIR-V operand kind: <a href="https://www.khronos.org/registry/spir-v/specs/unified1/SPIRV.html#_a_id_fp_denorm_mode_a_fp_denorm_mode">FPDenormMode</a></div></li><li><div class="item-name"><a class="enum" href="enum.FPOperationMode.html" title="enum spirv::FPOperationMode">FPOperationMode</a></div><div class="desc docblock-short">SPIR-V operand kind: <a href="https://www.khronos.org/registry/spir-v/specs/unified1/SPIRV.html#_a_id_fp_operation_mode_a_fp_operation_mode">FPOperationMode</a></div></li><li><div class="item-name"><a class="enum" href="enum.FPRoundingMode.html" title="enum spirv::FPRoundingMode">FPRoundingMode</a></div><div class="desc docblock-short">SPIR-V operand kind: <a href="https://www.khronos.org/registry/spir-v/specs/unified1/SPIRV.html#_a_id_fp_rounding_mode_a_fp_rounding_mode">FPRoundingMode</a></div></li><li><div class="item-name"><a class="enum" href="enum.FunctionParameterAttribute.html" title="enum spirv::FunctionParameterAttribute">FunctionParameterAttribute</a></div><div class="desc docblock-short">SPIR-V operand kind: <a href="https://www.khronos.org/registry/spir-v/specs/unified1/SPIRV.html#_a_id_function_parameter_attribute_a_function_parameter_attribute">FunctionParameterAttribute</a></div></li><li><div class="item-name"><a class="enum" href="enum.GLOp.html" title="enum spirv::GLOp">GLOp</a></div><div class="desc docblock-short"><a href="https://www.khronos.org/registry/spir-v/specs/unified1/GLSL.std.450.html">GLSL.std.450</a> extended instruction opcode</div></li><li><div class="item-name"><a class="enum" href="enum.GroupOperation.html" title="enum spirv::GroupOperation">GroupOperation</a></div><div class="desc docblock-short">SPIR-V operand kind: <a href="https://www.khronos.org/registry/spir-v/specs/unified1/SPIRV.html#_a_id_group_operation_a_group_operation">GroupOperation</a></div></li><li><div class="item-name"><a class="enum" href="enum.HostAccessQualifier.html" title="enum spirv::HostAccessQualifier">HostAccessQualifier</a></div><div class="desc docblock-short">SPIR-V operand kind: <a href="https://www.khronos.org/registry/spir-v/specs/unified1/SPIRV.html#_a_id_host_access_qualifier_a_host_access_qualifier">HostAccessQualifier</a></div></li><li><div class="item-name"><a class="enum" href="enum.ImageChannelDataType.html" title="enum spirv::ImageChannelDataType">ImageChannelDataType</a></div><div class="desc docblock-short">SPIR-V operand kind: <a href="https://www.khronos.org/registry/spir-v/specs/unified1/SPIRV.html#_a_id_image_channel_data_type_a_image_channel_data_type">ImageChannelDataType</a></div></li><li><div class="item-name"><a class="enum" href="enum.ImageChannelOrder.html" title="enum spirv::ImageChannelOrder">ImageChannelOrder</a></div><div class="desc docblock-short">SPIR-V operand kind: <a href="https://www.khronos.org/registry/spir-v/specs/unified1/SPIRV.html#_a_id_image_channel_order_a_image_channel_order">ImageChannelOrder</a></div></li><li><div class="item-name"><a class="enum" href="enum.ImageFormat.html" title="enum spirv::ImageFormat">ImageFormat</a></div><div class="desc docblock-short">SPIR-V operand kind: <a href="https://www.khronos.org/registry/spir-v/specs/unified1/SPIRV.html#_a_id_image_format_a_image_format">ImageFormat</a></div></li><li><div class="item-name"><a class="enum" href="enum.InitializationModeQualifier.html" title="enum spirv::InitializationModeQualifier">InitializationModeQualifier</a></div><div class="desc docblock-short">SPIR-V operand kind: <a href="https://www.khronos.org/registry/spir-v/specs/unified1/SPIRV.html#_a_id_initialization_mode_qualifier_a_initialization_mode_qualifier">InitializationModeQualifier</a></div></li><li><div class="item-name"><a class="enum" href="enum.KernelEnqueueFlags.html" title="enum spirv::KernelEnqueueFlags">KernelEnqueueFlags</a></div><div class="desc docblock-short">SPIR-V operand kind: <a href="https://www.khronos.org/registry/spir-v/specs/unified1/SPIRV.html#_a_id_kernel_enqueue_flags_a_kernel_enqueue_flags">KernelEnqueueFlags</a></div></li><li><div class="item-name"><a class="enum" href="enum.LinkageType.html" title="enum spirv::LinkageType">LinkageType</a></div><div class="desc docblock-short">SPIR-V operand kind: <a href="https://www.khronos.org/registry/spir-v/specs/unified1/SPIRV.html#_a_id_linkage_type_a_linkage_type">LinkageType</a></div></li><li><div class="item-name"><a class="enum" href="enum.LoadCacheControl.html" title="enum spirv::LoadCacheControl">LoadCacheControl</a></div><div class="desc docblock-short">SPIR-V operand kind: <a href="https://www.khronos.org/registry/spir-v/specs/unified1/SPIRV.html#_a_id_load_cache_control_a_load_cache_control">LoadCacheControl</a></div></li><li><div class="item-name"><a class="enum" href="enum.MemoryModel.html" title="enum spirv::MemoryModel">MemoryModel</a></div><div class="desc docblock-short">SPIR-V operand kind: <a href="https://www.khronos.org/registry/spir-v/specs/unified1/SPIRV.html#_a_id_memory_model_a_memory_model">MemoryModel</a></div></li><li><div class="item-name"><a class="enum" href="enum.Op.html" title="enum spirv::Op">Op</a></div><div class="desc docblock-short">SPIR-V <a href="https://www.khronos.org/registry/spir-v/specs/unified1/SPIRV.html#_a_id_instructions_a_instructions">instructions</a> opcodes</div></li><li><div class="item-name"><a class="enum" href="enum.OverflowModes.html" title="enum spirv::OverflowModes">OverflowModes</a></div><div class="desc docblock-short">SPIR-V operand kind: <a href="https://www.khronos.org/registry/spir-v/specs/unified1/SPIRV.html#_a_id_overflow_modes_a_overflow_modes">OverflowModes</a></div></li><li><div class="item-name"><a class="enum" href="enum.PackedVectorFormat.html" title="enum spirv::PackedVectorFormat">PackedVectorFormat</a></div><div class="desc docblock-short">SPIR-V operand kind: <a href="https://www.khronos.org/registry/spir-v/specs/unified1/SPIRV.html#_a_id_packed_vector_format_a_packed_vector_format">PackedVectorFormat</a></div></li><li><div class="item-name"><a class="enum" href="enum.QuantizationModes.html" title="enum spirv::QuantizationModes">QuantizationModes</a></div><div class="desc docblock-short">SPIR-V operand kind: <a href="https://www.khronos.org/registry/spir-v/specs/unified1/SPIRV.html#_a_id_quantization_modes_a_quantization_modes">QuantizationModes</a></div></li><li><div class="item-name"><a class="enum" href="enum.RayQueryCandidateIntersectionType.html" title="enum spirv::RayQueryCandidateIntersectionType">RayQueryCandidateIntersectionType</a></div><div class="desc docblock-short">SPIR-V operand kind: <a href="https://www.khronos.org/registry/spir-v/specs/unified1/SPIRV.html#_a_id_ray_query_candidate_intersection_type_a_ray_query_candidate_intersection_type">RayQueryCandidateIntersectionType</a></div></li><li><div class="item-name"><a class="enum" href="enum.RayQueryCommittedIntersectionType.html" title="enum spirv::RayQueryCommittedIntersectionType">RayQueryCommittedIntersectionType</a></div><div class="desc docblock-short">SPIR-V operand kind: <a href="https://www.khronos.org/registry/spir-v/specs/unified1/SPIRV.html#_a_id_ray_query_committed_intersection_type_a_ray_query_committed_intersection_type">RayQueryCommittedIntersectionType</a></div></li><li><div class="item-name"><a class="enum" href="enum.RayQueryIntersection.html" title="enum spirv::RayQueryIntersection">RayQueryIntersection</a></div><div class="desc docblock-short">SPIR-V operand kind: <a href="https://www.khronos.org/registry/spir-v/specs/unified1/SPIRV.html#_a_id_ray_query_intersection_a_ray_query_intersection">RayQueryIntersection</a></div></li><li><div class="item-name"><a class="enum" href="enum.SamplerAddressingMode.html" title="enum spirv::SamplerAddressingMode">SamplerAddressingMode</a></div><div class="desc docblock-short">SPIR-V operand kind: <a href="https://www.khronos.org/registry/spir-v/specs/unified1/SPIRV.html#_a_id_sampler_addressing_mode_a_sampler_addressing_mode">SamplerAddressingMode</a></div></li><li><div class="item-name"><a class="enum" href="enum.SamplerFilterMode.html" title="enum spirv::SamplerFilterMode">SamplerFilterMode</a></div><div class="desc docblock-short">SPIR-V operand kind: <a href="https://www.khronos.org/registry/spir-v/specs/unified1/SPIRV.html#_a_id_sampler_filter_mode_a_sampler_filter_mode">SamplerFilterMode</a></div></li><li><div class="item-name"><a class="enum" href="enum.Scope.html" title="enum spirv::Scope">Scope</a></div><div class="desc docblock-short">SPIR-V operand kind: <a href="https://www.khronos.org/registry/spir-v/specs/unified1/SPIRV.html#_a_id_scope_a_scope">Scope</a></div></li><li><div class="item-name"><a class="enum" href="enum.SourceLanguage.html" title="enum spirv::SourceLanguage">SourceLanguage</a></div><div class="desc docblock-short">SPIR-V operand kind: <a href="https://www.khronos.org/registry/spir-v/specs/unified1/SPIRV.html#_a_id_source_language_a_source_language">SourceLanguage</a></div></li><li><div class="item-name"><a class="enum" href="enum.StorageClass.html" title="enum spirv::StorageClass">StorageClass</a></div><div class="desc docblock-short">SPIR-V operand kind: <a href="https://www.khronos.org/registry/spir-v/specs/unified1/SPIRV.html#_a_id_storage_class_a_storage_class">StorageClass</a></div></li><li><div class="item-name"><a class="enum" href="enum.StoreCacheControl.html" title="enum spirv::StoreCacheControl">StoreCacheControl</a></div><div class="desc docblock-short">SPIR-V operand kind: <a href="https://www.khronos.org/registry/spir-v/specs/unified1/SPIRV.html#_a_id_store_cache_control_a_store_cache_control">StoreCacheControl</a></div></li></ul><h2 id="constants" class="section-header">Constants<a href="#constants" class="anchor">§</a></h2><ul class="item-table"><li><div class="item-name"><a class="constant" href="constant.MAGIC_NUMBER.html" title="constant spirv::MAGIC_NUMBER">MAGIC_NUMBER</a></div></li><li><div class="item-name"><a class="constant" href="constant.MAJOR_VERSION.html" title="constant spirv::MAJOR_VERSION">MAJOR_VERSION</a></div></li><li><div class="item-name"><a class="constant" href="constant.MINOR_VERSION.html" title="constant spirv::MINOR_VERSION">MINOR_VERSION</a></div></li><li><div class="item-name"><a class="constant" href="constant.REVISION.html" title="constant spirv::REVISION">REVISION</a></div></li></ul><h2 id="types" class="section-header">Type Aliases<a href="#types" class="anchor">§</a></h2><ul class="item-table"><li><div class="item-name"><a class="type" href="type.Word.html" title="type spirv::Word">Word</a></div></li></ul></section></div></main></body></html>