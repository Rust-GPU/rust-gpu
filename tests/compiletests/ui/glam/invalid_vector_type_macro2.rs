// build-fail
// normalize-stderr-test "\S*/crates/spirv-std/src/" -> "$$SPIRV_STD_SRC/"

use core::num::NonZeroU32;
use spirv_std::glam::Vec2;
use spirv_std::spirv;

#[spirv_std::spirv_vector]
#[derive(Copy, Clone, Default)]
pub struct ZstVector;

#[spirv_std::spirv_vector]
#[derive(Copy, Clone, Default)]
pub struct NotVectorField {
    _x: Vec2,
    _y: Vec2,
}

#[spirv_std::spirv_vector]
#[derive(Copy, Clone)]
pub struct NotVectorField2 {
    _x: NonZeroU32,
    _y: NonZeroU32,
}

impl Default for NotVectorField2 {
    fn default() -> Self {
        Self {
            _x: NonZeroU32::new(1).unwrap(),
            _y: NonZeroU32::new(1).unwrap(),
        }
    }
}

#[spirv(fragment)]
pub fn entry(
    // workaround to ZST loading
    #[spirv(storage_class, descriptor_set = 0, binding = 0)] _: &(ZstVector, i32),
    _: NotVectorField,
    #[spirv(flat)] _: NotVectorField2,
) {
}
