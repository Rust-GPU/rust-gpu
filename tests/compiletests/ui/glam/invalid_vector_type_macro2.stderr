error: Vector ZST not allowed
  --> $DIR/invalid_vector_type_macro2.rs:9:1
   |
LL | / #[derive(Copy, Clone, Default)]
LL | | pub struct ZstVector;
   | |_____________________^

error[E0412]: cannot find type `ZstVector` in this scope
  --> $DIR/invalid_vector_type_macro2.rs:38:67
   |
LL |     #[spirv(storage_class, descriptor_set = 0, binding = 0)] _: &(ZstVector, i32),
   |                                                                   ^^^^^^^^^ not found in this scope

error: unknown argument to spirv attribute
  --> $DIR/invalid_vector_type_macro2.rs:38:13
   |
LL |     #[spirv(storage_class, descriptor_set = 0, binding = 0)] _: &(ZstVector, i32),
   |             ^^^^^^^^^^^^^

error[E0277]: the trait bound `Vec2: Scalar` is not satisfied
  --> $DIR/invalid_vector_type_macro2.rs:15:9
   |
LL |     _x: Vec2,
   |         ^^^^ the trait `Scalar` is not implemented for `Vec2`
   |
   = help: the following other types implement trait `Scalar`:
             bool
             f32
             f64
             i16
             i32
             i64
             i8
             u16
           and 3 others
note: required by a bound in `spirv_std::ScalarOrVector::Scalar`
  --> $SPIRV_STD_SRC/scalar_or_vector.rs:18:18
   |
LL |     type Scalar: Scalar;
   |                  ^^^^^^ required by this bound in `ScalarOrVector::Scalar`

error[E0277]: the trait bound `Vec2: Scalar` is not satisfied
  --> $DIR/invalid_vector_type_macro2.rs:12:1
   |
LL | #[spirv_std::spirv_vector]
   | ^^^^^^^^^^^^^^^^^^^^^^^^^^ the trait `Scalar` is not implemented for `Vec2`
   |
   = help: the following other types implement trait `Scalar`:
             bool
             f32
             f64
             i16
             i32
             i64
             i8
             u16
           and 3 others
note: required by a bound in `Vector`
  --> $SPIRV_STD_SRC/vector.rs:56:28
   |
LL | pub unsafe trait Vector<T: Scalar, const N: usize>: ScalarOrVector<Scalar = T> {}
   |                            ^^^^^^ required by this bound in `Vector`
   = note: this error originates in the attribute macro `spirv_std::spirv_vector` (in Nightly builds, run with -Z macro-backtrace for more info)

error[E0277]: the trait bound `NonZero<u32>: Scalar` is not satisfied
  --> $DIR/invalid_vector_type_macro2.rs:22:9
   |
LL |     _x: NonZeroU32,
   |         ^^^^^^^^^^ the trait `Scalar` is not implemented for `NonZero<u32>`
   |
   = help: the following other types implement trait `Scalar`:
             bool
             f32
             f64
             i16
             i32
             i64
             i8
             u16
           and 3 others
note: required by a bound in `spirv_std::ScalarOrVector::Scalar`
  --> $SPIRV_STD_SRC/scalar_or_vector.rs:18:18
   |
LL |     type Scalar: Scalar;
   |                  ^^^^^^ required by this bound in `ScalarOrVector::Scalar`

error[E0277]: the trait bound `NonZero<u32>: Scalar` is not satisfied
  --> $DIR/invalid_vector_type_macro2.rs:19:1
   |
LL | #[spirv_std::spirv_vector]
   | ^^^^^^^^^^^^^^^^^^^^^^^^^^ the trait `Scalar` is not implemented for `NonZero<u32>`
   |
   = help: the following other types implement trait `Scalar`:
             bool
             f32
             f64
             i16
             i32
             i64
             i8
             u16
           and 3 others
note: required by a bound in `Vector`
  --> $SPIRV_STD_SRC/vector.rs:56:28
   |
LL | pub unsafe trait Vector<T: Scalar, const N: usize>: ScalarOrVector<Scalar = T> {}
   |                            ^^^^^^ required by this bound in `Vector`
   = note: this error originates in the attribute macro `spirv_std::spirv_vector` (in Nightly builds, run with -Z macro-backtrace for more info)

error: aborting due to 7 previous errors

Some errors have detailed explanations: E0277, E0412.
For more information about an error, try `rustc --explain E0277`.
